examples/kfpytorch_plugin/kfpytorch_plugin/pytorch_lightning_mnist_autoencoder.py

```diff
@@ -0,0 +1,151 @@
+        "s3fs==2024.3.1",
+    ],
+    cuda="12.1.0",
+    python_version="3.9",
+    registry="nielsbantilan",
+)
+
+# %% [markdown]
+# We're also going to define a custom pod template that mounts a shared memory
+# volume to `/dev/shm`. This is necessary for distributed data parallel (DDP)
+# training so that state can be shared across workers.
+
+container = V1Container(name=custom_image.name, volume_mounts=[V1VolumeMount(mount_path="/dev/shm", name="dshm")])
+volume = V1Volume(name="dshm", empty_dir=V1EmptyDirVolumeSource(medium="", size_limit="200Gi"))
+custom_pod_template = PodTemplate(
+    primary_container_name=custom_image.name,
+    pod_spec=V1PodSpec(containers=[container], volumes=[volume]),
+)
+
+# %% [markdown]
+# ## Define a Lightning Module
+#
+# Then we create a pytorch lightning module, which defines an autoencoder that
+# will learn how to create compressed embeddings of MNIST images.
+
+
+class LitAutoEncoder(L.LightningModule):
+    def __init__(self, encoder, decoder):
+        super().__init__()
+        self.encoder = encoder
+        self.decoder = decoder
+
+    def training_step(self, batch, batch_idx):
+        x, y = batch
+        x = x.view(x.size(0), -1)
+        z = self.encoder(x)
+        x_hat = self.decoder(z)
+        loss = nn.functional.mse_loss(x_hat, x)
+        self.log("train_loss", loss)
+        return loss
+
+    def configure_optimizers(self):
+        optimizer = optim.Adam(self.parameters(), lr=1e-3)
+        return optimizer
+
+
+# %% [markdown]
+# ## Creating the pytorch `Elastic` task
+#
+# With the model architecture defined, we now create a Flyte task that assumes
+# a world size of 16: 2 nodes with 8 devices each. We also set the `max_restarts`
+# to `3` so that the task can be retried up to 3 times in case it fails for
+# whatever reason, and we set `rdzv_configs` to have a generous timeout so that
+# the head and worker nodes have enought time to connect to each other.
+#
+# This task will output a `FlyteDirectory`, which will contain the model
+# checkpoint that will result from training.
+
+NUM_NODES = 2
+NUM_DEVICES = 8
+
+
+@task(
+    container_image=custom_image,
+    task_config=Elastic(
+        nnodes=NUM_NODES,
+        nproc_per_node=NUM_DEVICES,
+        rdzv_configs={"timeout": 36000, "join_timeout": 36000},
+        max_restarts=3,
+    ),
+    accelerator=T4,
+    requests=Resources(mem="32Gi", cpu="48", gpu="8", ephemeral_storage="100Gi"),
+    pod_template=custom_pod_template,
+)
+def train_model() -> FlyteDirectory:
+    encoder = nn.Sequential(nn.Linear(28 * 28, 64), nn.ReLU(), nn.Linear(64, 3))
+    decoder = nn.Sequential(nn.Linear(3, 64), nn.ReLU(), nn.Linear(64, 28 * 28))
+    autoencoder = LitAutoEncoder(encoder, decoder)
+
+    root_dir = os.getcwd()
+    dataset = MNIST(
+        os.path.join(root_dir, "dataset"),
+        download=True,
+        transform=ToTensor(),
+    )
+    train_loader = utils.data.DataLoader(dataset, num_workers=0)
```

Comment:
not sure, I can try?

Reply:
is this what you got?

```
"2024-05-07T20:42:10.273560419Z stderr F   File \"/opt/conda/envs/envd/lib/python3.9/multiprocessing/spawn.py[2024-05-07 20:42:10,273] torch.distributed.elastic.multiprocessing.api: [ERROR] failed (exitcode: -7) local_rank: 5 (pid: 3815) of fn: spawn_helper (start_method: spawn)"
2024-05-07T20:42:10.995535917Z stderr F /opt/conda/envs/envd/lib/python3.9/multiprocessing/resource_tracker.py:216: UserWarning: resource_tracker: There appear to be 266 leaked semaphore objects to clean up at shutdown"
```

Reply:
That error looks familiar when I was testing out lightning + Flyte.

In my case, I was trying to get it to work with one node. Does your workflow work if you remove the elastic configuration?

Reply:
for the record, managed to fix this issue by setting `medium="Memory"` in the volume definition:
```python
volume = V1Volume(name="dshm", empty_dir=V1EmptyDirVolumeSource(medium="Memory"))
```

instead of `medium=""`, which will use the default storage medium of the node.